Default config_path if legacy registry options are not set.

evankanderson opened the pull request:
Addresses https://github.com/containerd/containerd/issues/6485

evankanderson commented on pkg/cri/server/image_pull.go:
```diff
@@ -330,6 +330,12 @@ func hostDirFromRoots(roots []string) func(string) (string, error) {
+	if len(cR.Mirrors) == 0 && len(cR.Configs) == 0 && len(cR.Auths) == 0 && len(cR.Headers) == 0 {
+		// No old-style configuration, default to docker-compatible defaults
+		// See https://github.com/containerd/containerd/issues/6485
+		c.config.Registry.ConfigPath = "/etc/containerd/certs.d:/etc/docker/certs.d"
```
This is Linux-specific. I don't know if I need to do something cross-platform here or not.

kzys replied:
Do you know Docker's default in Windows? @jterry75 @kevpar

evankanderson replied:
Windows server directions are here; the TL;DR is that the certificates need to be installed into the system cert store on Windows, I think: https://docs.docker.com/registry/insecure/#use-self-signed-certificates

evankanderson replied:
In addition, it seems like the Windows cert directions require restarting Docker, so it's not as seamless as the Linux process of putting certificates under the registry name in `/etc/docker/certs.d`.

adisky replied:
would it help to set it here https://github.com/containerd/containerd/blob/main/pkg/cri/config/config_unix.go#L29?

adisky replied:
would it help to set it here https://github.com/containerd/containerd/blob/main/pkg/cri/config/config_unix.go#L29?

evankanderson replied:
Sorry, needed to do a little research.

Setting the default in `config_unix` causes the default config to be _merged_ with any config in the paths, which means that you can hit the [error condition on line 400 of `config.go`](https://github.com/containerd/containerd/blob/main/pkg/cri/config/config.go#L401) by setting any of the (deprecated) `mirrors` or `tls` parameters and not explicitly setting `config_path` to the empty string.

dmcgowan replied:
It is still best to catch the conflict there. This is kind of deep in the path to start applying default values. Where is it checking whether `ConfigPath` is already set?

evankanderson replied:
Good point!

I've moved this to the end of [`ValidatePluginConfig`](https://github.com/containerd/containerd/blob/main/pkg/cri/config/config.go) and updated the appropriate tests in `config_test.go`.

evankanderson commented:
(DCO-signed)

evankanderson commented:
/assign @cpuguy83

evankanderson commented:
/assign @kzys

kzys reviewed (approved):
Thanks @evankanderson. LGTM!

kzys commented:
/ok-to-test

adisky reviewed (commented):
Thanks @evankanderson LGTM
can you update the docs as well? 
https://github.com/containerd/containerd/blob/main/docs/cri/registry.md
https://github.com/containerd/containerd/blob/main/docs/cri/config.md

evankanderson commented:
Updated docs

adisky commented:
LGTM

evankanderson commented:
Is there anything else I need to do before this can be merged? (I don't have permissions to merge it myself.)

evankanderson commented on pkg/cri/config/config.go:
```diff
@@ -439,6 +444,17 @@ func ValidatePluginConfig(ctx context.Context, c *PluginConfig) error {
 		log.G(ctx).Warning("`auths` is deprecated, please use `configs` instead")
 	}
 
+	if len(c.Registry.Mirrors) == 0 && !hasDeprecatedTLS && len(c.Registry.Auths) == 0 {
```
Removed `Registry.Auths` here as well.

evankanderson commented on pkg/cri/config/config.go:
```diff
@@ -419,6 +421,9 @@ func ValidatePluginConfig(ctx context.Context, c *PluginConfig) error {
 	// Validation for deprecated auths options and mapping it to configs.
 	if len(c.Registry.Auths) != 0 {
+		if useConfigPath {
+			return errors.New("`auths` is deprecated, please use `config_path` instead")
```
The code looks like it supports setting an `authorizer`, but I've reverted this portion of the change and updated the tests.

evankanderson reviewed (commented):
Let me know if you want me to squash the commits before merging the PR.

evankanderson commented:
Anything else I need to do?

estesp commented:
Can you squash your commits? We have one LGTM from a containerd maintainer so will need at least one more before merging. I've marked it as ready for review since you had already made the updates @dmcgowan asked for.

evankanderson commented:
Squashed. Ready for another look.

mikebrow reviewed (commented):
thinking this over...

right now default for configpath is "" ..  (containerd config default > config.toml)

if a user edits the configpath in the config.toml the old mirror/tls has to have already been removed (checked by validator)

with this change we introduce a behavior change

if the configpath is "" and if mirrors/tls isn't set.. then nil configpath means use "/etc/containerd/certs.d:/etc/docker/certs.d"

an admin isn't informed or directing containerd to do that .. it just happens... default is now promoted to auto config from docker/certs.d + containerd/certs.d

if they don't want that to happen they have to add a dummy mirror..

this would be a behavior difference between 1.5 and main 

it should be noted that the default config does not have mirrors/tls in it

... 

my gut says it would be better, in main, to change the actual default value for config_path in new default configs from "" to the proposed, than to come up with a behavior change for what config_path = "" means retroactively against older configs...

evankanderson commented:
Sorry for the late reply; other things came up and then this got buried under more time-sensitive work.


> thinking this over...
> 
> right now default for configpath is "" .. (containerd config default > config.toml)
> 
> if a user edits the configpath in the config.toml the old mirror/tls has to have already been removed (checked by validator)
> 
> with this change we introduce a behavior change
> 
> if the configpath is "" and if mirrors/tls isn't set.. then nil configpath means use "/etc/containerd/certs.d:/etc/docker/certs.d"
> 
> an admin isn't informed or directing containerd to do that .. it just happens... default is now promoted to auto config from docker/certs.d + containerd/certs.d

That was the intent, mostly trying to follow two principles:

1. New and existing users should be able to use auto-config if possible. Auto-config has the following benefits over the existing configuration:
   * It doesn't require restarting `containerd` to pick up new certs.
   * The "syntax" to get things right is a bit easier than adding it to the TOML (at least in my experience).
   * It's simpler to add new files in new directories than to update an existing TOML file from shell commands.
3. Existing users who have 1.4 configurations shouldn't be broken on `containerd` upgrade (i.e. an `apt-get upgrade` shouldn't break existing configurations)

> if they don't want that to happen they have to add a dummy mirror..

Or simply not create the directories / delete them, or point the `configPath` to something like `/dev/null`. I think I'm slightly less concerned about the auto-configuration being turned on because it requires multiple actions to activate:

2. Create a directory at one of the default locations (all of which should be root-owned / restricted write).
3. Create a _second_ directory matching a registry name underneath one of the directory locations.
4. Add the appropriate files inside the second directory.

> this would be a behavior difference between 1.5 and main
>
> it should be noted that the default config does not have mirrors/tls in it

I added this conditionally because it appeared that the default configuration and the toml file were merged. If the two are merged, then you can end up with a config file which looks like:

```toml
[plugins."io.containerd.grpc.v1.cri".registry.configs."registry.mycorp.com".tls]
ca_file   = "/etc/docker/certs.d/registry.mycorp.com/ca.crt"
```

and after the merge with defaults it looks like:

```toml
[plugins."io.containerd.grpc.v1.cri".registry]
config_path = "/etc/containerd/certs.d:/etc/docker/certs.d"
[plugins."io.containerd.grpc.v1.cri".registry.configs."registry.mycorp.com".tls]
ca_file   = "/etc/docker/certs.d/registry.mycorp.com/ca.crt"
```

Which will fail validation, because legacy `ca_file` and `configPath` are set at the same time. This would violate my principle "existing users who have 1.4 configurations shouldn't be broken on `containerd` upgrade". To make this valid, you'd need to explicitly set `config_path` to `""` in your config before upgrading to 1.7.

Trying to back out the setting of `configPath` if any of the legacy options are set seemed equally risky. I *could* start ignoring the legacy options in 1.7 if you wanted, but that seemed like more of a backwards compatibility break than you might want.

I agree, the current situation kind of sucks -- suggestions welcome!

> my gut says it would be better, in main, to change the actual default value for config_path in new default configs from "" to the proposed, than to come up with a behavior change for what config_path = "" means retroactively against older configs...

I've switched to your suggestion as an additional commit; let me know what I can do to help with testing.

> and whatever is needed for config_windows.go

Windows works totally differently for Docker et al and uses the windows cert store, rather than directories.

evankanderson commented:
(I can clean up the PR history if we agree we want to change this in `config_unix.go` after the above discussion.)

evankanderson commented:
/retest

kzys commented:
/retest

evankanderson commented:
I think the failures are from [following @mikebrow 's suggestion](https://github.com/containerd/containerd/pull/6488#pullrequestreview-902309825) and making the registry config part of the default configuration, but the logs don't include the `journalctl` output from containerd, so I can't tell if we're [hitting the assertion on old-and-new registry configuration I mentioned in my previous comment](https://github.com/containerd/containerd/pull/6488#issuecomment-1100709575).

evankanderson commented:
@mikebrow -- see my comments above about defaulting the setting like one would wish.

If we allowed both the `config.toml` options and the directory path to co-exist (with one having precedence), I think it might be possible to merge the default config and the user's existing config, but that would be a change over the current design direction.

evankanderson commented:
Ping @mikebrow -- I know you wanted this changed in the defaults, but [this comment](https://github.com/containerd/containerd/pull/6488#issuecomment-1100709575) explains why I don't think that's going to work (that's the current test-failing state of the current code).

I'd love to get this into the next `containerd` release to enable us to start a 2 or 3 year clock ticking on providers using a more up-to-date version with this option defaulted.

evankanderson commented:
Ping again, @mikebrow -- did my comments make sense? Do you have a different solution for clearing out new-style options if old-style ones are parsed?

sagor999 commented:
I would love to see this merged as well, as we currently dealing with the exact same problem, of having to modify containerd config so that it can actually read CA cert for our docker registry.
It would be great if that would have been a default behavior.

evankanderson commented:
@mikebrow -- any thoughts on my comments in https://github.com/containerd/containerd/pull/6488#issuecomment-1100709575 about changing the default configuration and the difficulty of determining "new defaults" or "conflicting configuration"?

cpuguy83 commented:
We could probably wait for 2.0 for this (should be release after 1.7).

mikebrow commented:
agree with Brian that we could do this prior to 2.0 as a larger refactor of the auth config...

evankanderson commented:
I'm willing to take another swing at this if someone wants to spec out the
behavior they want to see.

Alternatively, I can write up a spec for the first version of this PR and
we can go from there.

On Tue, Nov 1, 2022, 2:38 PM Mike Brown ***@***.***> wrote:

> agree with Brian that we could do this prior to 2.0 as a larger refactor
> of the auth config...
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/containerd/containerd/pull/6488#issuecomment-1299234381>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB4XEN2KVT5IAVTLLAYQDO3WGGEULANCNFSM5M7LLQ7Q>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

ruiwen-zhao commented:
I have seen some of our customers wanting exactly the same to happen, so would like to see this merged.

To summarize, prior to 2.0, we would like to 
1. change the default config_path to as proposed in this PR, and
2. handling the conflict between config_path and the legacy `mirrors` and/or `configs.tls` , if set by users. And we haven't yet agreed on how to handle the conflict.

Did I summarize it correctly?

If so may I propose that in 2.0 we just ignore the lagacy  `mirrors` and `configs.tls` (which might have been the plan any way)?

evankanderson commented:
FWIW, I'm willing to work on updating this and preparing it for a merge if
I can get feedback that it's likely to be accepted.

If this is only going into 2.0, I'm less concerned about but breaking
existing customers who have the legacy configuration, and can do the
simpler thing Mike suggested.

On Fri, Feb 10, 2023, 1:49 PM Ruiwen Zhao ***@***.***> wrote:

> I have seen some of our customers wanting exactly the same to happen, so
> would like to see this merged.
>
> To summarize, prior to 2.0, we would like to
>
>    1. change the default config_path to as proposed in this PR, and
>    2. handling the conflict between config_path and the legacy mirrors
>    and/or configs.tls , if set by users. And we haven't yet agreed on how
>    to handle the conflict.
>
> Did I summarize it correctly?
>
> If so may I propose that in 2.0 we just ignore the lagacy mirrors and
> configs.tls (which might have been the plan any way)?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/containerd/containerd/pull/6488#issuecomment-1426385635>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB4XEN4HS4FBKGV7G3OHAL3WW2ZXPANCNFSM5M7LLQ7Q>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

cpuguy83 commented:
So yes, we should be able to to just remove those from the configs whole-sale for 2.0 and move to this new config.

evankanderson commented:
I just noticed that it looks like this would also simplify https://github.com/kubernetes-sigs/kind/issues/2875.  Is there interest in my reviving this?  If so, what's the desired behavior that we'd want to see in the following cases:

1. Existing configuration which sets other `registry` options for specific hosts ("legacy configuration")
2. Existing configuration which did not set any `registry` options in an existing install
3. New installations without any existing configuration

adisky commented:
@evankanderson considering `plugins."io.containerd.grpc.v1.cri".registry.configs` are going away with 2.0, do you want to update this PR? This is the only chance to have default `config_path` set.

cc @mikebrow

mikebrow commented:
@dmcgowan FYI

mikebrow commented:
@dmcgowan fyi

evankanderson commented:
(I'm willing to revive this if there's interest, but it seemed like there wasn't much maintainer interest the last time I asked.)

evankanderson commented:
Is this a change in defaults we'd like to make for 2.0?

Per discussion on the Kubernetes lists, it sounds like the 2.0 release may be just around the corner, and breaking existing users (in rare cases) seemed like the main concern with this change.

AkihiroSuda commented:
Looks good, but please squash the commits.
Also consider explaining the background in the description of the PR to get more attention from maintainers.

AkihiroSuda commented:
I'm moving this to the v2.1 milestone, but I guess this can be safely merged to v2.0 too if we want.

mikebrow reviewed (approved):
looks good .. needs rebase

evankanderson commented:
/test pull-containerd-k8s-e2e-ec2

estesp commented:
@evankanderson looks like there is an actual test failure based on expectation for the old behavior:
```
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -51,3 +51,3 @@
        	            	     [plugins.'io.containerd.cri.v1.images'.registry]
        	            	-      config_path = '/etc/containerd/certs.d:/etc/docker/certs.d'
        	            	+      config_path = ''
        	            	 
        	Test:       	TestMigration/1.7-Default
    --- FAIL: TestMigration/1.7-Default (0.03s)

=== FAIL: integration/client TestMigration (0.11s)
```

This is affecting most of the Linux CI runs

evankanderson commented:
> @evankanderson looks like there is an actual test failure based on expectation for the old behavior:
> 
> ```
>         	            	Diff:
>         	            	--- Expected
>         	            	+++ Actual
>         	            	@@ -51,3 +51,3 @@
>         	            	     [plugins.'io.containerd.cri.v1.images'.registry]
>         	            	-      config_path = '/etc/containerd/certs.d:/etc/docker/certs.d'
>         	            	+      config_path = ''
>         	            	 
>         	Test:       	TestMigration/1.7-Default
>     --- FAIL: TestMigration/1.7-Default (0.03s)
> 
> === FAIL: integration/client TestMigration (0.11s)
> ```
> 
> This is affecting most of the Linux CI runs

Thanks!  I tried doing the update last night on my Windows machine, and I don't seem to have the same test environment as the tests anyway.

I added a migration to `plugins/cri/images/plugin.go`, because that _looked_ like the right way to handle the migration tests.  But I'm not familiar with this code, so please double-check me.

/no-idea-dog

evankanderson commented:
/retest

Romain-Geissler-1A commented:
Hi,

What is the status on this change ? Is the failing test the blocking point ? If yes, does anyone has an idea how to fix it ?

mxpv commented:
/retest

dmcgowan commented:
Rebased to run another round of CI on latest main

evankanderson commented:
I'm willing to do some bounded additional amount of work here, but I'm not sure that I'll be able to debug the current issues, as I don't understand the test harnesses and environments well enough to recognize flakes vs failures.

evankanderson commented on internal/cri/config/config_unix.go:
I don't know enough about Windows (or how to test) to suggest anything here.
